lib/salestation/rspec.rb

```diff
@@ -0,0 +1,47 @@
+# frozen_string_literal: true
+
+require 'rspec/expectations'
+require 'salestation/app'
+
+require_relative './rspec/failure_matcher'
+require_relative './rspec/glia_input_validation_error_matcher'
+
+module Salestation
+  module RSpec
+    # Usage:
+    #
+    # In your RSpec configuration first include the matchers:
+    #
+    #   RSpec.configure do |config|
+    #     config.include Salestation::RSpec::Matchers
+    #   end
+    #
+    # Then you can use the matchers like this:
+    #
+    #   expect(result).to be_failure
+    #     .with_conflict
+    #     .containing(hash_including(message: 'Oh noes'))
+    #
+    # or when using Glia Errors:
+    #
+    #   expect(result).to be_failure
+    #     .with_requested_resource_not_found
+    #     .based_on(Glia::Errors::ResourceNotFoundError.new(resource: :user))
```

Comment:
I wonder if we actually need `based_on`.
For new endpoints (without legacy fields) it should be possible to have 

```
expect(result).to be_failure
  .with_conflict
  .containing(Glia::Errors::ResourceNotFoundError.new(resource: :user).to_h)
```

for response that have legacy fields could be:

```
expect(result).to be_failure
  .with_conflict
  .containing(Glia::Errors::ResourceNotFoundError.new(resource: :user).to_h.merge(message: ''))
```

or 

```
expect(result).to be_failure
  .with_conflict
  .containing(hash_including(Glia::Errors::ResourceNotFoundError.new(resource: :user).to_h))
```

and 

```
expect(result).to be_failure
  .with_invalid_input
  .contains(glia_input_validation_error.on(:name).with_type(Glia::Errors::INVALID_VALUE_ERROR))
```

Reply:
The app errors have a separate field for `base_error`. So it would need to be something like:
```
  .containing(hash_including(base_error: Glia::Errors::ResourceNotFoundError.new(resource: :user).to_h))
```

Though I think I could rewrite it a bit and make an exception. So if it detects an object then it treats as a base error object.

Btw, currently if you want to do both base_error and legacy you can also do:
```
expect(result).to be_failure
  .with_conflict
  .based_on(Glia::Errors::ResourceNotFoundError.new(resource: :user))
  .containing(hash_including(message: 'foobar'))
```

But maybe it makes sense to default to `hash_including` and add an exception for objects. So it would look like:
```
expect(result).to be_failure
  .with_conflict
  .containing(Glia::Errors::ResourceNotFoundError.new(resource: :user))
  .containing(message: 'foobar')
```

Reply:
Or if the message field is very common then could also support: `  .with_message('foobar')`

Reply:
https://github.com/salemove/salestation/compare/d851c6136ae594961f2c65372640762435d7ca0d..e51b07d36ba9caec4e985c085eedc7a11bc506a4 `containing` not also accepts the base error. Removed `based_on` function.

Reply:
the `based_on` method still seems to be present and used once https://github.com/salemove/salestation/pull/51/commits/cea7e00af16e1cd2c4f9f91a9119ba3e7bf3f3f8#diff-0951ac3b1911df4126eeb138e7392124b709062546b6f4f9f6f977ae4c697313R49

Reply:
fixed https://github.com/salemove/salestation/compare/a1843e7f2c11f1d51cdb053beed8fd7ede202d49..ec77ef1eb98adf3faa263952c28fef975c01be35